presto-main/src/main/java/com/facebook/presto/cost/SimpleFilterProjectSemiJoinStatsRule.java

```diff
@@ -138,22 +167,52 @@ else if (nodeSource instanceof SemiJoinNode) {
                 .filter(conjunct -> conjunct != semiJoinOutputReference)
                 .collect(toImmutableList()));
         boolean negated = semiJoinOutputReference instanceof NotExpression;
+        return Optional.of(new SemiJoinOutputFilter(negated, castToRowExpression(remainingPredicate)));
+    }
+
+    private Optional<SemiJoinOutputFilter> extractSemiJoinOutputFilter(RowExpression predicate, RowExpression input)
+    {
+        checkState(!isExpression(predicate));
+        List<RowExpression> conjuncts = LogicalRowExpressions.extractConjuncts(predicate);
+        List<RowExpression> semiJoinOutputReferences = conjuncts.stream()
+                .filter(conjunct -> isSemiJoinOutputReference(conjunct, input))
+                .collect(toImmutableList());
+
+        if (semiJoinOutputReferences.size() != 1) {
+            return Optional.empty();
+        }
+
+        RowExpression semiJoinOutputReference = Iterables.getOnlyElement(semiJoinOutputReferences);
+        RowExpression remainingPredicate = logicalRowExpressions.combineConjuncts(conjuncts.stream()
+                .filter(conjunct -> conjunct != semiJoinOutputReference)
+                .collect(toImmutableList()));
+        boolean negated = isNegated(semiJoinOutputReference);
         return Optional.of(new SemiJoinOutputFilter(negated, remainingPredicate));
     }
 
+    private boolean isSemiJoinOutputReference(RowExpression conjunct, RowExpression input)
+    {
+        return conjunct.equals(input) || (isNegated(conjunct) && ((CallExpression) conjunct).getArguments().get(0).equals(input));
+    }
+
     private static boolean isSemiJoinOutputReference(Expression conjunct, Symbol semiJoinOutput)
     {
         SymbolReference semiJoinOuputSymbolReference = semiJoinOutput.toSymbolReference();
         return conjunct.equals(semiJoinOuputSymbolReference) ||
                 (conjunct instanceof NotExpression && ((NotExpression) conjunct).getValue().equals(semiJoinOuputSymbolReference));
     }
 
+    private boolean isNegated(RowExpression expression)
```

Comment:
This name is quite similar to the `NEGATION` operator. It might be better to stick with `isNotFunction`.